### PR TITLE
ST4: 4065+ HTMLSheet Updates

### LINKED ISSUE #94: Not showing properly, HTML elements being displayed in preview
image

As you can see above, the preview is showing the HTML elements of the markdown conversion. For reference, this is my MD format:

```bash
# Counting
ls | wc -l                                                                              # Count files in a directory (including directories)
find -type f | sed -e 's/.*\.//' | sort | uniq -c                                       # Count files in directory per each type of extension
ls | grep '\.jpg$' | wc -l                                                              # Count all files in a directory with one particular extension

# Listing
ls | grep '\.jpg$'                                                                      # List all files with a particular extension
find -type f -size +50000k -exec ls -lh {} \; | awk '{ print $9 ": " $5 }'              # Find all files in a directory with a specific size value (50000k or larger)
du --max-depth=1 | sort -n | awk 'BEGIN {OFMT = "%.0f"} {print $1/1024,"MB", $2}'       # List all folders in a directory sorted by MB

# Compare
diff <(cd folder1 && find . | sort) <(cd folder2 && find . | sort)                      # Compare the contents fo two folders

# Bulk Jobs
rename 's/\.htm$/\.html/' *.htm                                                         # Bulk rename all files in a folder (from .htm to .html)
rename -n 's/\.htm$/\.html/' *.htm                                                      # Test run, doesn't actually change but shows proposed changes

# Backup 
`sudo -u advadmin cp -p filename newfilename.txt.YYYYMMDD`                              # Create a file backup

# Search
grep -rl "looking_for*" /path/                                                          # Search all files in directory for "string*" and list the files
```

When I build the MD into an HTML file it displays properly, it's just live view that doesn't work.

Using MarkdownEditing plugin, Macbook Pro 2019 MacOS Mojave, Sublime 3.2.2 build 3211.

### LINKED ISSUE #98: Unable to preview the markdown.
No option to `MarkdownLivePreview: Open Preview` were found and even pressing `alt+m` didn't work

### LINKED ISSUE #103: Images not loaded correctly.
Schermata 2020-02-27 alle 17 31 12

### LINKED ISSUE #104: Support for language specific code blocks.
It's a lovely tool.
It would be nice to have support for blocks like 
<pre>
```python
print("Hi")
```
</pre>
too.  At the moment this gives some strange output: 

Screen Shot 2020-03-01 at 8 03 59 pm

Screen Shot 2020-03-01 at 8 04 01 pm

### LINKED ISSUE #106: Some code is  display error
image

### PR DESCRIPTION
**Pull request by TerminalFi**
**Should be a ST4 only branch**
- Requires => `4074`

Will update the Package Control Channel to support ST3 and ST4 tags and set appropriate ST4 version requirement

### Major Changes

- [x] Removes `2markdown.py` in favor of Mdpopups
  - Provides proper code highlighting and handles md2html
- [x] Removes `markdown2html.py` implements `ImageParser.py`
  - Utilizes 95% of the same code with minor tweaks
- [x] Adds `add_on_change` for monitoring the `SETTING_DELAY_BETWEEN_UPDATES`
to update the delay in real time. (**Could be back ported**)
- [x] Adds `MdlpEraseCommand` for erasing a view when called from `on_pre_close` **_Should probably be implemented on the ST3 branch to fix a bug in which EventListener doesn't have a `edit` object_**
https://github.com/math2001/MarkdownLivePreview/blob/ec27d980a3ebd4c73c23f9036e5b4edf1dea2e27/MarkdownLivePreview.py#L130-L135
- [x] Replaces Phantoms with HTMLSheets which are implemented in ST4065
- [x] Implements a `PREVIEW_VIEWS` hash map to allow for multiple markdown previews (**Could be back ported**)
- [x] Adds a `syntax` setting that allows you to enable the plugin on different syntax files (**Could be back ported**)
- [ ] #22

### Fixes

fixes #94 - Fixed with ST4 Changes. Likely can fix in ST3 by using Mdpopups but not HTMLSheet
fixes #98 via facelessuser/sublime-markdown-popups@5c65d6f
fixes #103
fixes #104
fixes #106

### CODE REVIEW

**Comment by jrappen** [comment]
@TheSecEng: please add to the above message

---

```text
fixes #98 via facelessuser/sublime-markdown-popups@5c65d6f
fixes #103
fixes #104
```

**Comment by TheSecEng** [comment]
Investigating an issue with `mdpopups` where it doesn't convert some Markdown correctly

File: MarkdownLivePreview.py [hunk 1]
```diff
@@ -78,11 +141,8 @@ def run(self, edit):
         )
 
         preview_window.focus_group(1)
-        preview_view = preview_window.new_file()
-        preview_view.set_scratch(True)
-        preview_view.settings().set(PREVIEW_VIEW_INFOS, {})
-        preview_view.set_name("Preview")
-        # FIXME: hide number lines on preview
+
+        preview_view = mdpopups.new_html_sheet(preview_window, "Preview", "")
```
Comment (jrappen) [inline MarkdownLivePreview.py]: I suggest changing this to:

`preview_view = mdpopups.new_html_sheet(window=preview_window, name="Preview", contents="")`

File: MarkdownLivePreview.py [hunk 2]
```diff
@@ -169,72 +218,76 @@ def on_close(self, markdown_view):
...
-                    lambda href: sublime.run_command("open_url", {"url": href}),
-                )
-            ]
+        file_name = view.file_name()
+        basepath = os.path.dirname(file_name) if file_name else None
+        html_content = imageparser(
+            html_content, basepath, partial(self._update_preview, view), resources,
         )
 
+        mdpopups.update_html_sheet(preview_view, html_content, md=False)
```
Comment (jrappen) [inline MarkdownLivePreview.py]: I suggest changing this to:

`mdpopups.update_html_sheet(sheet=preview_view, contents=html_content, md=False)`

**Comment by jrappen** [comment]
ST Build 4074 and mdpopups `v3.7.0` had breaking changes due to Sublime no longer supporting `cmd` and `args` for `sublime.Window.new_html_sheet`, at least breaking the `4065-4073` behavior.

Code is fine as is, but requirement should be bumped.

File: ImageParser.py [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+
+__all__ = ("imageparser",)
+
+
+RE_BAD_ENTITIES = re.compile(r"(&(?!amp;|lt;|gt;|nbsp;)(?:\w+;|#\d+;))")
+
+# FIXME: how do I choose how many workers I want?
+# - Does thread pool reuse threads or does it stupidly throw them out?
+# - (we could implement something of our own)
+executor = concurrent.futures.ThreadPoolExecutor(max_workers=5)
```
Comment (rwols) [inline ImageParser.py:18]: This is overkill for this sort of plugin.

File: ImageParser.py [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+        content_type = conn.info().get_content_type()
+        if "image" not in content_type:
+            raise ValueError(
+                "{!r} doesn't point to an image, but to a {!r}".format(
+                    url, content_type
+                )
+            )
+        return "{}{}".format(
+            "data:image/png;base64,", b64encode(image_content).decode("utf-8")
+        )
```
Comment (rwols) [inline ImageParser.py:135]: Looks like you want to use asyncio instead of a ThreadPoolExecutor (which is IMO completely the wrong tool for this job).

**Comment by AmjadHD** [comment]
Why did work on this stopped ?

**Comment by TerminalFi** [comment]
> Why did work on this stopped ?



> Why did work on this stopped ?

I just am in the middle of moving and what not. I’d be happy to get help with it. 

I’ll start working on it again when I have time.

**Pull request closed without merging**